### PR TITLE
make the host_blocker aware of patterns

### PR DESCRIPTION
**Pull request by Ongy**
Add an additional argument to is_blocked for the current url, and enable
the host_blocking.enabled property to be set per domain.
This allows to bypass host blocking for specific domains

---
This change is [Reviewable](https://reviewable.io/reviews/qutebrowser/qutebrowser/4046)

### CODE REVIEW

File: qutebrowser/browser/adblock.py [hunk 1]
```diff
@@ -111,9 +111,9 @@ def __init__(self):
 
         config.instance.changed.connect(self._update_files)
 
-    def is_blocked(self, url):
+    def is_blocked(self, url, primary=None):
         """Check if the given URL (as QUrl) is blocked."""
-        if not config.val.content.host_blocking.enabled:
+        if not config.instance.get("content.host_blocking.enabled", url=primary):
```
Comment (The-Compiler) [inline qutebrowser/browser/adblock.py]: Nitpick: Please use single quotes here (I try to use single quotes for "keys" like this, and double-quotes for "actual" strings shown to the user).

File: qutebrowser/browser/adblock.py [hunk 2]
```diff
@@ -111,9 +111,9 @@ def __init__(self):
 
         config.instance.changed.connect(self._update_files)
 
-    def is_blocked(self, url):
+    def is_blocked(self, url, primary=None):
```
Comment (The-Compiler) [inline qutebrowser/browser/adblock.py]: Not sure `primary` is a good name for the argument - why not `first_party_url` to be consistent with what QtWebEngine uses?

File: qutebrowser/browser/adblock.py (hunk 1 above)
Comment (The-Compiler) [reply qutebrowser/browser/adblock.py]: You'll also need to split this into two lines (e.g. after the comma)

File: qutebrowser/browser/webkit/network/networkmanager.py [hunk 3]
```diff
@@ -408,6 +400,14 @@ def createRequest(self, op, req, outgoing_data):
                 # the webpage shutdown here.
                 current_url = QUrl()
```
Comment (jgkamat) [inline qutebrowser/browser/webkit/network/networkmanager.py]: Are `QUrl()` and `None` equivalent for per domain settings? If there is, there will be a descrepancy between this and when the method is called with missing args (if it is in the future).

Comment (Ongy) [reply qutebrowser/browser/webkit/network/networkmanager.py]: Ufff, you are asking the right question.
I hope someone knows^^

Comment (toofar) [reply qutebrowser/browser/webkit/network/networkmanager.py]: Good catch, it looks like they [are treated differently](https://github.com/Ongy/qutebrowser/blob/40b2d03688ab14703fe8eae6f8d98e578e623995/qutebrowser/config/configutils.py#L158) and an empty `QUrl` is apparently invalid so a `QtValueError` will be thrown from [here](https://github.com/Ongy/qutebrowser/blob/40b2d03688ab14703fe8eae6f8d98e578e623995/qutebrowser/utils/urlmatch.py#L278).

Comment (The-Compiler) [reply qutebrowser/browser/webkit/network/networkmanager.py]: Good catch! Fixed in c6dc9206e6cdedc74a401fc2827c1f855b945805.

**Comment by The-Compiler** [comment]
Thanks, and sorry for the delay! I also added a test in fd48e29a54f839c4538367aeea549f8def8706a7.

**Pull request merged**